### PR TITLE
Update assembly version from hardcoded to MajorVersion and update NetCoreAppCurrent to net8.0

### LINKED ISSUE #58109: Tests which depend on the host don't use a live built version
Both the linker tests and the singlefile tests use a prebuilt version of the host instead of the live built one. In general we want to use live built assets where possible. In addition to that, when the repository revs to the next major version and the SDK doesn't yet understand the net7.0 tfm, the prebuilt host isn't discovered without help (`KnownAppHostPack` AppHostPackVersion update).

TL;DR: Tests which depend on the apphost should have a dependency on the host subset being built first.

### PR DESCRIPTION
**Pull request by carlossanlop**
Contributes to https://github.com/dotnet/runtime/issues/72112
Fixes https://github.com/dotnet/runtime/issues/58109

Co-authored with @radical and @ViktorHofer 

Original PR was https://github.com/dotnet/runtime/pull/74157 but we had to revert it https://github.com/dotnet/runtime/pull/77899

Reason why it had to be reverted:

> _Changing assembly versions without changing the TargetFramework represented by those assemblies is causing any consuming assembly in packages that target net7.0 to have references to 8.0 versioned assemblies. This will break the consumption of those packages._

> _For this change to work, it cannot change referenced versions that are exposed as net7.0. The TFM targeted by packages can likely change before AssemblyVersion, but not vice-versa._

### CODE REVIEW

File: Directory.Build.props
Comment (carlossanlop) [inline Directory.Build.props:49]: Shouldn't we be using the values we have in eng/Versions.props that we added in this PR? https://github.com/dotnet/runtime/pull/78354/commits/17764b999e7dfece2e72ca1bf7c68aa73465c8cd

Here, and also elsewhere where we hardcode the minor or major numbers.

File: src/libraries/System.Data.Odbc/src/System.Data.Odbc.csproj
Comment (carlossanlop) [inline src/libraries/System.Data.Odbc/src/System.Data.Odbc.csproj:3]: Fun! 😄

File: Directory.Build.props
Comment (carlossanlop) [inline Directory.Build.props:60]: Is this correct? This would become a `7.0.0`

Comment (ViktorHofer) [reply Directory.Build.props:60]: Yes, ApiCompat restores the previous stable package from nuget, which now is 7.0.0.

File: Directory.Build.props
Comment (ViktorHofer) [reply Directory.Build.props:49]: Directory.Build.props has access to Versions.props but we don't want to use the version properties from there as that would complicate branding to .NET 9, as it would mean that the package version branding change would also rev the NetCoreAppCurrent version, and we don't want that.

**Comment by carlossanlop** [comment]
@ViktorHofer do we need any other changes to prevent breaking the other repos, or are we good now?

I can't approve this PR because I submitted it, and I assume @ViktorHofer can't either since he contributed to it too.

@agocke / @ericstj / @smasher164 / @radical can we get a sign-off from at least one of you?

**Comment by ViktorHofer** [comment]
@carlossanlop we just started working on this yesterday and we have another meeting tomorrow in which we will continue to make progress. This PR requires many more changes (I expect another 50-100 file changes). Also, the build is failing because we haven't yet updated all the places that require updates.

**Comment by ViktorHofer** [comment]
> /__w/1/s/.dotnet/sdk/7.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(267,5): error NETSDK1005: Assets file '/__w/1/s/artifacts/tests/coreclr/packages/Common/test_dependencies/test_dependencies/project.assets.json' doesn't have a target for 'net8.0'. Ensure that restore has run and that you have included 'net8.0' in the TargetFrameworks for your project. [/__w/1/s/src/tests/JIT/Regression/JitBlue/Runtime_60035/Runtime_60035.csproj] [/__w/1/s/src/tests/build.proj]
##[error].dotnet/sdk/7.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(267,5): error NETSDK1005: Assets file '/__w/1/s/artifacts/tests/coreclr/packages/Common/test_dependencies/test_dependencies/project.assets.json' doesn't have a target for 'net8.0'. Ensure that restore has run and that you have included 'net8.0' in the TargetFrameworks for your project. [/__w/1/s/src/tests/JIT/Regression/JitBlue/Runtime_60035/Runtime_60035.csproj]

@jkoritzinsky @hoyosjs @agocke can you please look into the failure here? Should test_dependencies.csproj/fsproj target `$(NetCoreAppToolCurrent)` or `$(NetCoreAppCurrent)`? This PR updates `$(NetCoreAppCurrent)` to `net8.0`.

**Comment by ViktorHofer** [comment]
> C:\Program Files\dotnet\sdk\7.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1183: Unable to optimize assemblies for Ahead of time compilation: a valid runtime package was not found. Either set the PublishAot property to false, or use a supported runtime identifier when publishing. When targeting .NET 7 or higher, make sure to restore packages with the PublishAot property set to true. [D:\a\_work\1\s\artifacts\bin\trimmingTests\projects\System.Diagnostics.DiagnosticSource.NativeAotTests\DiagnosticSourceEventSourceTests\win-x64\project.csproj]

@MichalStrehovsky do you know who can help with that error? This PR upgrades the tfm from net7.0 to net8.0 and I'm seeing the linker tests now failing. That's from the `windows x64 release NativeAOT` leg.

**Comment by agocke** [comment]
I’m looking at the test dependencies. @sbomer  can you look at the Native aot error? Might be the same warning problem

**Comment by radical** [comment]
- wasm/debugger tests should be fixed now.
- And the failing wasi build on `LibraryTests` job should work too

**Comment by radical** [comment]
Wasm/debugger-tests/windows will be fixed by https://github.com/dotnet/runtime/pull/79124 .

Wasm threading, and perforating builds are [failing with](https://dev.azure.com/dnceng-public/public/_build/results?buildId=99596&view=logs&jobId=1567862a-6a9e-5016-9353-cc84f1411dc4&j=8d0bfc3e-d1bc-5879-4cc4-18653f9adb55&t=d60309cd-b93c-533c-9d7c-31b2b5b95034):

```
  API compatibility errors between 'ref/net8.0/System.Diagnostics.Tracing.dll' (left) and 'lib/net8.0/System.Diagnostics.Tracing.dll' (right):
error CP0014: Cannot remove attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")' from 'System.Diagnostics.Tracing.DiagnosticCounter'. [/__w/1/s/src/libraries/System.Diagnostics.Tracing/src/System.Diagnostics.Tracing.csproj]
error CP0014: Cannot remove attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")' from 'System.Diagnostics.Tracing.EventCounter'. [/__w/1/s/src/libraries/System.Diagnostics.Tracing/src/System.Diagnostics.Tracing.csproj]
error CP0014: Cannot remove attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")' from 'System.Diagnostics.Tracing.IncrementingEventCounter'. [/__w/1/s/src/libraries/System.Diagnostics.Tracing/src/System.Diagnostics.Tracing.csproj]
error CP0014: Cannot remove attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")' from 'System.Diagnostics.Tracing.IncrementingPollingCounter'. [/__w/1/s/src/libraries/System.Diagnostics.Tracing/src/System.Diagnostics.Tracing.csproj]
error CP0014: Cannot remove attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")' from 'System.Diagnostics.Tracing.PollingCounter'. [/__w/1/s/src/libraries/System.Diagnostics.Tracing/src/System.Diagnostics.Tracing.csproj]
```

cc @pavelsavara @lambdageek

**Comment by radical** [comment]
And I pushed a change to fix the wasm library tests build.

**Comment by ViktorHofer** [comment]
> A note about the workaround: I tried to work around the missing KnownILCompilerPack by instead adding it in targetingpacks.targets, and then removing items from PackageDownload. This wasn't working because the nativeaot tests aren't restored in a standard way:

Thanks for providing the workaround @sbomer. I just tried your original approach out and it worked for me. I removed your workaround and added a `KnownILCompilerPack` entry instead. One general note, now that the libraries linker and aot tests use a life apphost, I had to add the `clr.runtime+host.native` subset to the nativeaot build lanes. I just wanted to build the 
apphost via the `host.native` subset but that didn't work without building the singlefilehost first. I opened https://github.com/dotnet/runtime/issues/79144 for that.

**Comment by ViktorHofer** [comment]
Who could help with the DNNE error? I initially switched it to use the live app host but that resulted in a dependency on the host subset when building the libraries tests which we want to avoid (with our current loosely coupled subset system). Now it again uses the SDK host package that should be available, but I have no idea why it's failing to compile.

```
  Building native export: "/usr/bin/clang-9" -O2 -shared -fpic -D DNNE_ASSEMBLY_NAME=Microsoft.Interop.Tests.NativeExports -D DNNE_COMPILE_AS_SOURCE -I "/__w/1/s/.packages/dnne/1.0.27/tools/platform" -I "/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native" -o "/__w/1/s/artifacts/obj/NativeExports/Release/net8.0/linux-musl-arm/dnne/bin/Microsoft.Interop.Tests.NativeExportsNE.so" --target=armv7-alpine-linux-musleabihf --gcc-toolchain=/crossrootfs/arm/usr --sysroot=/crossrootfs/arm -mthumb -mfpu=vfpv3 -DCLR_ARM_FPU_CAPABILITY=0x7 "/__w/1/s/artifacts/obj/NativeExports/Release/net8.0/linux-musl-arm/dnne/Microsoft.Interop.Tests.NativeExports.g.c" "/__w/1/s/.packages/dnne/1.0.27/tools/platform/platform.c" -lstdc++ "/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native/libnethost.a" --target=armv7-alpine-linux-musleabihf --gcc-toolchain=/crossrootfs/arm/usr --sysroot=/crossrootfs/arm -fuse-ld=lld  -Wl,--rpath-link=/crossrootfs/arm/lib/armv7-alpine-linux-musleabihf -Wl,--rpath-link=/crossrootfs/arm/usr/lib/armv7-alpine-linux-musleabihf 
clang : error : no such file or directory: '/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native/libnethost.a' [/__w/1/s/src/libraries/System.Runtime.InteropServices/tests/TestAssets/NativeExports/NativeExports.csproj]
##[error]clang(0,0): error : (NETCORE_ENGINEERING_TELEMETRY=Build) no such file or directory: '/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native/libnethost.a'
/__w/1/s/.packages/dnne/1.0.27/build/DNNE.targets(145,5): error MSB3073: The command ""/usr/bin/clang-9" -O2 -shared -fpic -D DNNE_ASSEMBLY_NAME=Microsoft.Interop.Tests.NativeExports -D DNNE_COMPILE_AS_SOURCE -I "/__w/1/s/.packages/dnne/1.0.27/tools/platform" -I "/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native" -o "/__w/1/s/artifacts/obj/NativeExports/Release/net8.0/linux-musl-arm/dnne/bin/Microsoft.Interop.Tests.NativeExportsNE.so" --target=armv7-alpine-linux-musleabihf --gcc-toolchain=/crossrootfs/arm/usr --sysroot=/crossrootfs/arm -mthumb -mfpu=vfpv3 -DCLR_ARM_FPU_CAPABILITY=0x7 "/__w/1/s/artifacts/obj/NativeExports/Release/net8.0/linux-musl-arm/dnne/Microsoft.Interop.Tests.NativeExports.g.c" "/__w/1/s/.packages/dnne/1.0.27/tools/platform/platform.c" -lstdc++ "/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native/libnethost.a" --target=armv7-alpine-linux-musleabihf --gcc-toolchain=/crossrootfs/arm/usr --sysroot=/crossrootfs/arm -fuse-ld=lld  -Wl,--rpath-link=/crossrootfs/arm/lib/armv7-alpine-linux-musleabihf -Wl,--rpath-link=/crossrootfs/arm/usr/lib/armv7-alpine-linux-musleabihf " exited with code 1. [/__w/1/s/src/libraries/System.Runtime.InteropServices/tests/TestAssets/NativeExports/NativeExports.csproj]
##[error].packages/dnne/1.0.27/build/DNNE.targets(145,5): error MSB3073: (NETCORE_ENGINEERING_TELEMETRY=Build) The command ""/usr/bin/clang-9" -O2 -shared -fpic -D DNNE_ASSEMBLY_NAME=Microsoft.Interop.Tests.NativeExports -D DNNE_COMPILE_AS_SOURCE -I "/__w/1/s/.packages/dnne/1.0.27/tools/platform" -I "/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native" -o "/__w/1/s/artifacts/obj/NativeExports/Release/net8.0/linux-musl-arm/dnne/bin/Microsoft.Interop.Tests.NativeExportsNE.so" --target=armv7-alpine-linux-musleabihf --gcc-toolchain=/crossrootfs/arm/usr --sysroot=/crossrootfs/arm -mthumb -mfpu=vfpv3 -DCLR_ARM_FPU_CAPABILITY=0x7 "/__w/1/s/artifacts/obj/NativeExports/Release/net8.0/linux-musl-arm/dnne/Microsoft.Interop.Tests.NativeExports.g.c" "/__w/1/s/.packages/dnne/1.0.27/tools/platform/platform.c" -lstdc++ "/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native/libnethost.a" --target=armv7-alpine-linux-musleabihf --gcc-toolchain=/crossrootfs/arm/usr --sysroot=/crossrootfs/arm -fuse-ld=lld  -Wl,--rpath-link=/crossrootfs/arm/lib/armv7-alpine-linux-musleabihf -Wl,--rpath-link=/crossrootfs/arm/usr/lib/armv7-alpine-linux-musleabihf " exited with code 1.
```

cc @AaronRobinsonMSFT

**Comment by ViktorHofer** [comment]
The installer tests are failing because of https://github.com/dotnet/runtime/issues/45972.

**Comment by AaronRobinsonMSFT** [comment]
> no such file or directory: '/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native/libnethost.a

This seems like the root issue. Is this file there? Is the path to the musl lib different in this case? I've no idea why this would be failing, but do have an arm64 macOS machine so will try to repro locally.

**Comment by AaronRobinsonMSFT** [comment]
Starting from a clean repo (i.e., `git clean -xdf`) on an arm64-macOS I ran the following command:

`./build.sh -subset libs+libs.tests   -configuration Debug -arch arm64 -framework net8.0 -testscope innerloop`

This is the same command in the above failing macOS leg. I am thinking this is a race condition with the file system and restore because this works on other legs without issue and macOS nor linux is special cased in any way with respect to command execution. I am going to re-run the failing legs again to test this hypothesis.

The MSBuild target in DNNE for this area can be found at https://github.com/AaronRobinsonMSFT/DNNE/blob/236974815bbefd80d844d9c38f8fb154df33684b/src/msbuild/DNNE.targets#L106-L157. It is very simple and I've no idea how anything in there could be causing the failure only on musl and arm64-macos.

**Comment by ViktorHofer** [comment]
> This is the same command in the above failing macOS leg. I am thinking this is a race condition with the file system and restore because this works on other legs without issue and macOS nor linux is special cased in any way with respect to command execution. I am going to re-run the failing legs again to test this hypothesis.

I think this is more about legs that cross-build, as only such fail on CI. Except for FreeBSD but apparently NativeExports isn't compiled on that leg at all.

**Review by ViktorHofer — APPROVED** [review]
Approving others and my own changes.

**Comment by AaronRobinsonMSFT** [comment]
@ViktorHofer Can I do anything in DNNE to make this better?

**Comment by agocke** [comment]
> > A note about the workaround: I tried to work around the missing KnownILCompilerPack by instead adding it in targetingpacks.targets, and then removing items from PackageDownload. This wasn't working because the nativeaot tests aren't restored in a standard way:
> 
> Thanks for providing the workaround @sbomer. I just tried your original approach out and it worked for me. I removed your workaround and added a `KnownILCompilerPack` entry instead. Related, now that the libraries linker and aot tests use the live apphost, I had to add the `clr.runtime+host.native` subset to the nativeaot build lanes. I just wanted to build the apphost via the `host.native` subset but that didn't work without building the singlefilehost first. I opened #79144 for that.

I don't want to block this PR, but consider just redefining `clr.aot` to be whatever you had to write to get this to work. The `clr.aot` partition is meant to be "whatever's needed for AOT in common cases", not a specific set of stuff.

**Review by agocke — APPROVED** [review]

**Comment by ViktorHofer** [comment]
> I don't want to block this PR, but consider just redefining clr.aot to be whatever you had to write to get this to work. The clr.aot partition is meant to be "whatever's needed for AOT in common cases", not a specific set of stuff.

That sounds like a good strategy until we have [strongly defined dependencies in the repo](https://github.com/dotnet/runtime/issues/43110). But before adding the `clr.runtime` subset to `clr.aot` we should first resolve the discussion in https://github.com/dotnet/runtime/issues/79144 as I don't think that the `clr.runtime` subset is necessary at all but makes the AOT build significantly slower.

**Comment by ViktorHofer** [comment]
> @ViktorHofer Can I do anything in DNNE to make this better?

DNNE's msbuild integration is well written and allows to customize any of the used properties. I doubt that we would need improvements there to simplify our build. The problem is dotnet/runtime itself which makes it hard to depend on live built artifacts (like the host) because of the way it currently builds in CI (runtime and libraries in parallel on separate machines).

**Comment by MichalStrehovsky** [comment]
> That sounds like a good strategy until we have https://github.com/dotnet/runtime/issues/43110. But before adding the clr.runtime subset to clr.aot we should first resolve the discussion in https://github.com/dotnet/runtime/issues/79144 as I don't think that the clr.runtime subset is necessary at all and makes the AOT build significantly slower.

Agreed, especially if the host is only needed to run the `RunNativeAotTestApps` mode introduced in #76109. This is not a set of tests a runtime developer would run - they guard libraries invariants and are not interesting for runtime devs. I really wish we did something so that the host is not looked for in that mode (not in this PR I guess). I don't know why it's looked for.

File: docs/workflow/building/coreclr/nativeaot.md
Comment (MichalStrehovsky) [inline docs/workflow/building/coreclr/nativeaot.md:104]: Is this really necessary?

I thought the failures we saw were related to the `RunNativeAotTestApps` step that was added in #76109. We don't run that in this doc.

Comment (ViktorHofer) [reply docs/workflow/building/coreclr/nativeaot.md:104]: Thanks, you are right. I mixed up `RunNativeAotTestApps` and `TestNativeAot`. Will send a follow-up to revert this change to not block this PR.

**Pull request merged**
